### ISSUE TITLE
002-ADR-Syncing-Strategy

### ISSUE THREAD

**Issue by Shyam-Sundar-Bharathi** [Shyam-Sundar-Bharathi/cse210-fa24-group14#4]

**Comment by Shyam-Sundar-Bharathi** [comment]
# ADR-003: Syncing Strategy

## Context and Problem Statement
Our browser extension project requires a data persistence and syncing strategy that supports storing user data (e.g., journal entries, paired URLs) and syncing it across multiple devices for a consistent user experience. Given our goal of keeping the project lightweight and minimizing complexity, we must decide how to store and synchronize data effectively without using a backend server or server-side infrastructure.

## Decision Drivers
1. **Simplicity**: We want to avoid the complexity and overhead of managing a backend server, which would require infrastructure setup, security measures, and ongoing maintenance.
2. **Lightweight Architecture**: A serverless approach aligns with our project’s aim to be simple, fast, and easy to deploy.
3. **User Experience**: Users should have a seamless experience, with their data accessible and up-to-date across devices. The chosen strategy should work well with the browser environment.
4. **Reliability and Performance**: Data should be stored and synced efficiently without impacting the performance of the extension or the user’s browser.
5. **Privacy and Security**: Data should be stored securely, adhering to best practices for protecting user information, especially when syncing across devices.

## Considered Options
### 1. **Local Storage and Browser Sync (Chosen Strategy)**
   - **Description**: Use the browser’s built-in cache (e.g., `localStorage` or `IndexedDB`) to store data locally on the user’s device. Leverage browser-provided syncing APIs (e.g., Chrome’s `chrome.storage.sync`) for syncing across devices.
   - **Pros**:
     - **No Server Infrastructure**: Eliminates the need for backend setup, reducing complexity and cost.
     - **Seamless Syncing**: Most modern browsers offer built-in syncing capabilities, allowing data to be shared across devices as long as the user is signed in with their browser account.
     - **Efficient Data Storage**: A browser cache ensures fast data access and minimal latency.
     - **Lightweight and Fast**: The extension remains lightweight and easy to install, as all data management is handled locally or through the browser’s sync features.
     - **Privacy by Design**: Data remains on the user’s devices or the browser's sync service, minimizing exposure to external servers and reducing privacy concerns.
   - **Cons**:
     - **Limited Storage Quota**: Browser sync APIs have storage limits (e.g., Chrome’s `storage.sync` has a quota of 100KB per item and 8KB per key), which may restrict the data we can store.
     - **Sync Delays**: Syncing data across devices may experience delays or be inconsistent, depending on the browser’s sync service reliability.
     - **Browser Dependency**: Syncing depends on the user’s browser and may not work as expected if they use multiple browsers or browsers that do not support the sync API.

### 2. **Local Storage with Manual Export/Import**
   - **Pros**:
     - **Full User Control**: Users can manually export their data to a file and import it on another device, providing flexibility and control.
     - **No Sync Limitations**: Avoids browser sync quotas or delays issues.
   - **Cons**:
     - **Poor User Experience**: Manual syncing is cumbersome and inconvenient, especially for frequently switching devices.
     - **Error-Prone**: Manual processes are prone to user error and could lead to data loss if not handled carefully.

### 3. **Backend Server for Data Syncing**
   - **Pros**:
     - **Unlimited Storage**: A backend server could handle larger data volumes without the restrictions imposed by browser sync APIs.
     - **Custom Sync Logic**: Allows fine-grained control over syncing and managing data.
   - **Cons**:
     - **Complexity and Cost**: Managing a backend server introduces significant complexity, including infrastructure setup, security, and maintenance.
     - **Privacy and Security Concerns**: Storing user data on a server increases the risk of data breaches and requires implementing strict security measures.
     - **Resource Overhead**: Adds latency and potential performance issues compared to local storage.

## Decision
We have chosen to use **Local Storage and Browser Sync** as our syncing strategy. Data will be stored locally using browser cache (e.g., `localStorage` or `IndexedDB`), ensuring fast and efficient access. We will use the browser’s built-in sync API for syncing across devices, such as `chrome.storage.sync`. This approach allows us to maintain a lightweight and efficient extension while providing a reasonably good user experience for data syncing.

## Consequences
### Positive Consequences
- **Simplified Architecture**: No need to set up or maintain backend servers, reducing development and operational overhead.
- **Cost-Effective**: Eliminates costs associated with hosting and maintaining a backend, making the project easier to sustain long-term.
- **Lightweight and Responsive**: The extension remains performant, as data is accessed directly from the browser cache.
- **Seamless Syncing**: Users benefit from automatic data synchronization across devices, provided they are signed into their browser account.
- **Data Privacy**: User data remains within the browser ecosystem, reducing the risk of exposure to third parties.

### Negative Consequences
- **Storage Limitations**: The browser sync API’s storage quotas may constrain the amount of data we can sync, requiring careful management and optimization of stored data.
- **Browser Dependency**: Syncing is tied to the capabilities and reliability of the browser’s sync service, which may vary across browsers or fail if users do not use the same browser ecosystem.
- **Limited Offline Capabilities**: Although local storage allows for offline access, syncing data while offline will not be possible until the device reconnects to the internet.

### Trade-Offs
- **Simplicity vs. Scalability**: We traded off the scalability and flexibility of a backend server for the simplicity and speed of a serverless approach. This is acceptable given our current project requirements but may need to be revisited if the scope or data needs to expand.
- **Sync Reliability vs. User Control**: By relying on browser sync, we lose some control over syncing behavior but gain a more seamless experience than manual data management.

## Related Architectural Decisions
- **Data Structure Optimization**: Given the storage limitations of browser sync, we must design our data structures to be as efficient as possible, using techniques like data compression or minimizing redundant data.

## Notes and Future Considerations
- **Data Quota Management**: We may need to implement data pruning strategies or offer users options to manage and clean up old data if we approach storage limits.
- **Alternative Sync Solutions**: If user feedback indicates issues with browser sync reliability or storage limits, we could explore hybrid solutions, like optional cloud storage for power users.
- **Cross-Browser Support**: As the project evolves, we should test and optimize for sync behavior across different browsers and consider browser-specific limitations and features.

This decision ensures that our browser extension remains efficient and user-friendly while avoiding the complexity of backend infrastructure. We will monitor and adapt based on user needs and feedback.

**Comment by kenny-lfx** [comment]
I completely agree with this syncing strategy. Using local storage with browser sync keeps the project lightweight, simple, and cost-effective while offering a decent user experience. The trade-off with storage limits and browser dependency seems reasonable, given the goal of avoiding backend complexity. Thanks for the great design!

**Comment by philemonputra** [comment]
This provides a robust foundation for a simple approach to data persistence and syncing in a browser extension, aligning well with the project’s simplicity. Adding more specific strategies around data management, cross-browser compatibility, and error handling could further strengthen this decision as the project scales. Overall, this approach is likely to support an efficient and responsive user experience without the complexity of backend infrastructure.

**Comment by shouhardik** [comment]
Local Storage and Browser Sync and is a good choosen option over the others given its simple architechure, lightweight, security and automatic sync across all devices. Additionally we get to use the built in APIs to streamline the process of storing and retrieving data.
